### PR TITLE
HADOOP-16965 Refactor abfs stream configuration.

### PR DESCRIPTION
**Pull request by mukund-thakur**
Ran all abfs tests using mvn -T 1C -Dparallel-tests=abfs clean verify with my storage account in us-east.

### CODE REVIEW

**Comment by mukund-thakur** [comment]
@steveloughran Does this makes sense?

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 12s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  22m  3s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 28s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 20s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 31s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  16m 28s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 23s |  trunk passed  |
| +0 :ok: |  spotbugs  |   1m  1s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 58s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 27s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 23s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 23s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 14s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 26s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 51s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 25s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   1m  9s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 10s |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 29s |  The patch does not generate ASF License warnings.  |
|  |   |  64m 43s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.8 Server=19.03.8 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1956 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux e831e2886726 4.15.0-91-generic #92-Ubuntu SMP Fri Feb 28 11:09:48 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / aeeebc5 |
| Default Java | 1.8.0_242 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/1/testReport/ |
| Max. process+thread count | 312 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   2m 36s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  1s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  26m 10s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 33s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 25s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 36s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  19m  7s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 25s |  trunk passed  |
| +0 :ok: |  spotbugs  |   1m  0s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 57s |  trunk passed  |
| -0 :warning: |  patch  |   1m 20s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 32s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 29s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 29s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 19s |  hadoop-tools/hadoop-azure: The patch generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
| +1 :green_heart: |  mvnsite  |   0m 32s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  18m 45s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 27s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   1m  9s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 35s |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 38s |  The patch does not generate ASF License warnings.  |
|  |   |  77m  4s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.8 Server=19.03.8 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1956 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 0cb8fdcf24e5 4.15.0-74-generic #84-Ubuntu SMP Thu Dec 19 08:06:28 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 4db598e |
| Default Java | 1.8.0_242 |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/2/artifact/out/diff-checkstyle-hadoop-tools_hadoop-azure.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/2/testReport/ |
| Max. process+thread count | 309 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/2/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.11.1 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsInputStream.java [hunk 1]
```diff
@@ -61,21 +61,19 @@
...
-      final long contentLength,
-      final int bufferSize,
-      final int readAheadQueueDepth,
-      final boolean tolerateOobAppends,
-      final String eTag) {
+          final AbfsClient client,
+          final Statistics statistics,
+          final String path,
+          final long contentLength,
+          AbfsInputStreamContext abfsInputStreamContext,
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsInputStream.java:68]: for consistency, mark as `final`. I know, it's not that significant in a constructor, but it just keeps things the same.

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStreamContext.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+  public AbfsOutputStreamContext disableOutputStreamFlush(
+          final boolean disableOutputStreamFlush) {
+    this.disableOutputStreamFlush = disableOutputStreamFlush;
+    return this;
+  }
+
+  public AbfsOutputStreamContext build() {
+    // Validation of parameters to be done here.
+    return this;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStreamContext.java:54]: do we need any validators here yet, or is there enough in the code as it is?

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsInputStreamContext.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  public AbfsInputStreamContext withReadBufferSize(final int readBufferSize) {
+    this.readBufferSize = readBufferSize;
+    return this;
+  }
+
+  public AbfsInputStreamContext withReadAheadQueueDepth(
+          final int readAheadQueueDepth) {
+    this.readAheadQueueDepth = (readAheadQueueDepth >= 0)
+            ? readAheadQueueDepth
+            : Runtime.getRuntime().availableProcessors();;
```
Comment (steveloughran) [inline hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsInputStreamContext.java:44]: trailing ;

File: hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStreamContext.java (hunk 2 above)
Comment (mukund-thakur) [reply hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsOutputStreamContext.java:54]: Now we don't have any validation yet but we can add the new ones here in future if required.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   0m 32s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| -1 :x: |  test4tests  |   0m  0s |  The patch doesn't appear to include any new or modified tests. Please justify why no new tests are needed for this patch. Also please list what manual steps were performed to verify this patch.  |
||| _ trunk Compile Tests _ |
| +1 :green_heart: |  mvninstall  |  20m 59s |  trunk passed  |
| +1 :green_heart: |  compile  |   0m 30s |  trunk passed  |
| +1 :green_heart: |  checkstyle  |   0m 22s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   0m 34s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  15m 15s |  branch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 27s |  trunk passed  |
| +0 :ok: |  spotbugs  |   0m 54s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   0m 52s |  trunk passed  |
| -0 :warning: |  patch  |   1m 12s |  Used diff version of patch file. Binary files and potentially other changes not applied. Please rebase and squash commits if necessary.  |
||| _ Patch Compile Tests _ |
| +1 :green_heart: |  mvninstall  |   0m 29s |  the patch passed  |
| +1 :green_heart: |  compile  |   0m 23s |  the patch passed  |
| +1 :green_heart: |  javac  |   0m 23s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 17s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   0m 27s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 25s |  patch has no errors when building and testing our client artifacts.  |
| +1 :green_heart: |  javadoc  |   0m 23s |  the patch passed  |
| +1 :green_heart: |  findbugs  |   0m 54s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 23s |  hadoop-azure in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 32s |  The patch does not generate ASF License warnings.  |
|  |   |  61m 21s |   |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/3/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1956 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux 00a114c67f04 4.15.0-58-generic #64-Ubuntu SMP Tue Aug 6 11:12:41 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 60fa153 |
| Default Java | Private Build-1.8.0_242-8u242-b08-0ubuntu3~16.04-b08 |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/3/testReport/ |
| Max. process+thread count | 414 (vs. ulimit of 5500) |
| modules | C: hadoop-tools/hadoop-azure U: hadoop-tools/hadoop-azure |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1956/3/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Pull request merged**